Table sync fail with ambiguous error (so far localised to 1 user)

### Describe the bug

Sync for a specific customer are failing

### To Reproduce

Not able to reproduce but happens to 1 customer on our cloud. More context here:

https://metaboat.slack.com/archives/C05MPF0TM3L/p1749654265453379

We're trying to remove as many tables as we can but the error seems to point to a Select * from the `metabase_table` and metabase_field` which I believe won't matter how many tables we remove from their Database. The problem is in our App DB

### Expected behavior

The sync works or errors with a proper message

### Logs

2025-06-11 13:14:38.381	
{"exception":{"exception_class":"clojure.lang.ExceptionInfo","stacktrace":"clojure.lang.ExceptionInfo: class clojure.lang.ExceptionInfo cannot be cast to class java.lang.Number (clojure.lang.ExceptionInfo is in unnamed module of loader 'app'; java.lang.Number is in module java.base of loader 'bootstrap') {:toucan2/context-trace [[\"execute SQL with class com.mchange.v2.c3p0.impl.NewProxyConnection\" {:toucan2.jdbc.query/sql-args [\"SELECT * FROM \\\"metabase_table\\\" WHERE ((\\\"active\\\" = TRUE) AND (\\\"visibility_type\\\" IS NULL)) AND (\\\"db_id\\\" = ?)\" 11]}] [\"resolve connection\" {:toucan2.connection/connectable metabase.app_db.connection.ApplicationDB}] [\"resolve connection\" {:toucan2.connection/connectable :default}] [\"resolve connection\" {:toucan2.connection/connectable nil}] {:toucan2.pipeline/rf #object[clojure.core$map$fn__5950$fn__5951 0x46c6a71a \"clojure.core$map$fn__5950$fn__5951@46c6a71a\"]} [\"with compiled query\" {:toucan2.pipeline/compiled-query [\"SELECT * FROM \\\"metabase_table\\\" WHERE ((\\\"active\\\" = TRUE) AND (\\\"visibility_type\\\" IS NULL)) AND (\\\"db_id\\\" = ?)\" 11]}] [\"with built query\" {:toucan2.pipeline/built-query {:select [:*], :from [[:metabase_table]], :where [:and [:and [:= :active true] [:= :visibility_type nil]] nil nil [:= :db_id 11]]}}] [\"with resolved query\" {:toucan2.pipeline/resolved-query {:where [:and [:and [:= :active true] [:= :visibility_type nil]] nil nil]}}] [\"with parsed args\" {:toucan2.pipeline/query-type :toucan.query-type/select.instances, :toucan2.pipeline/parsed-args {:kv-args {:db_id 11}, :queryable {:where [:and [:and [:= :active true] [:= :visibility_type nil]] nil nil]}}}] [\"with model\" {:toucan2.pipeline/model :model/Table}] [\"with unparsed args\" {:toucan2.pipeline/query-type :toucan.query-type/select.instances, :toucan2.pipeline/unparsed-args (:model/Table :db_id 11 {:where [:and [:and [:= :active true] [:= :visibility_type nil]] nil nil]})}]]}\n\tat metabase.task_history.models.task_history$do_with_task_history65145__65146.invokeStatic(task_history.clj:157)\n\tat metabase.task_history.models.task_history$do_with_task_history65145__65146.invoke(task_history.clj:145)\n\tat metabase.sync.util$run_sync_operation66063__66064.invokeStatic(util.clj:589)\n\tat metabase.sync.util$run_sync_operation66063__66064.invoke(util.clj:584)\n\tat metabase.sync.analyze$analyze_db_BANG_98326__98327$fn__98328.invoke(analyze.clj:116)\n\tat metabase.sync.util$do_with_error_handling.invokeStatic(util.clj:198)\n\tat metabase.sync.util$do_with_error_handling.invoke(util.clj:191)\n\tat clojure.core$partial$fn__5929.invoke(core.clj:2647)\n\tat metabase.driver$fn__53365.invokeStatic(driver.clj:896)\n\tat metabase.driver$fn__53365.invoke(driver.clj:896)\n\tat clojure.lang.MultiFn.invoke(MultiFn.java:239)\n\tat metabase.sync.util$sync_in_context$fn__65915.invoke(util.clj:170)\n\tat metabase.sync.util$with_db_logging_disabled$fn__65912.invoke(util.clj:162)\n\tat metabase.sync.util$with_start_and_finish_logging_STAR_.invokeStatic(util.clj:134)\n\tat metabase.sync.util$with_start_and_finish_logging_STAR_.invoke(util.clj:128)\n\tat metabase.sync.util$with_start_and_finish_logging$fn__65899.invoke(util.clj:147)\n\tat metabase.sync.util$with_sync_events65889__65890$fn__65894.invoke(util.clj:120)\n\tat metabase.sync.util$with_duplicate_ops_prevented$fn__65879.invoke(util.clj:92)\n\tat metabase.sync.util$do_sync_operation65938__65939.invokeStatic(util.clj:241)\n\tat metabase.sync.util$do_sync_operation65938__65939.invoke(util.clj:234)\n\tat metabase.sync.analyze$analyze_db_BANG_98326__98327.invokeStatic(analyze.clj:114)\n\tat metabase.sync.analyze$analyze_db_BANG_98326__98327.invoke(analyze.clj:109)\n\tat metabase.warehouses.api$fn__123701$fn__123707$fn__123713$fn__123714.invoke(api.clj:1023)\n\tat metabase_enterprise.database_routing.common$fn__120787$with_database_routing_off_fn__120788.invoke(common.clj:90)\n\tat clojure.lang.AFn.applyToHelper(AFn.java:154)\n\tat clojure.lang.AFn.applyTo(AFn.java:144)\n\tat clojure.core$apply.invokeStatic(core.clj:667)\n\tat clojure.core$apply.invoke(core.clj:662)\n\tat metabase.premium_features.defenterprise$dynamic_ee_oss_fn$fn__48307.doInvoke(defenterprise.clj:54)\n\tat clojure.lang.RestFn.invoke(RestFn.java:411)\n\tat metabase.warehouses.api$fn__123701$fn__123707$fn__123713.invoke(api.clj:1021)\n\tat clojure.lang.AFn.applyToHelper(AFn.java:152)\n\tat clojure.lang.AFn.applyTo(AFn.java:144)\n\tat clojure.core$apply.invokeStatic(core.clj:667)\n\tat clojure.core$with_bindings_STAR_.invokeStatic(core.clj:1990)\n\tat clojure.core$with_bindings_STAR_.doInvoke(core.clj:1990)\n\tat clojure.lang.RestFn.invoke(RestFn.java:428)\n\tat clojure.lang.AFn.applyToHelper(AFn.java:156)\n\tat clojure.lang.RestFn.applyTo(RestFn.java:135)\n\tat clojure.core$apply.invokeStatic(core.clj:671)\n\tat clojure.core$bound_fn_STAR_$fn__5837.doInvoke(core.clj:2020)\n\tat clojure.lang.RestFn.invoke(RestFn.java:400)\n\tat clojure.lang.AFn.call(AFn.java:18)\n\tat java.base/java.util.concurrent.FutureTask.run(Unknown Source)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)\n\tat java.base/java.lang.Thread.run(Unknown Source)\nCaused by: java.lang.ClassCastException: class clojure.lang.ExceptionInfo cannot be cast to class java.lang.Number (clojure.lang.ExceptionInfo is in unnamed module of loader 'app'; java.lang.Number is in module java.base of loader 'bootstrap')\n\t... 178 more\n","exception_message":"class clojure.lang.ExceptionInfo cannot be cast to class java.lang.Number (clojure.lang.ExceptionInfo is in unnamed module of loader 'app'; java.lang.Number is in module java.base of loader 'bootstrap')"},"m":"Error running step ''fingerprint-fields'' for snowflake Database 11 ''CBH Snowflake -- USE ME''","lgr":"metabase.sync.util","lvl":"WARN"}

### Information about your Metabase installation

```JSON
1.55.2 but used to happen to them before that version and why they migrated to cloud
```

### Severity

They aren't able to add any new tables to their metabase instance

### Additional context

Setting it as P2 since so far happens only to 1 user that I am aware of

--- a/src/metabase/sync/analyze/fingerprint.clj
+++ b/src/metabase/sync/analyze/fingerprint.clj
@@ -229,7 +229,7 @@
                                                (update :failed-fingerprints inc))
                                            ret)]
                                  (merge-with + acc ret))]
-                   (if (and (continue? ret) (not (sync-util/abandon-sync? ret)))
+                   (if (and (continue? new-acc) (not (sync-util/abandon-sync? ret)))
                      new-acc
                      (reduced new-acc))))
                (empty-stats-map 0)

--- a/src/metabase/sync/analyze/fingerprint.clj
+++ b/src/metabase/sync/analyze/fingerprint.clj
@@ -222,8 +222,14 @@
                     (sync-util/reducible-sync-tables database))]
        (reduce (fn [acc table]
                  (log-progress-fn (if *refingerprint?* "refingerprint-fields" "fingerprint-fields") table)
-                 (let [new-acc (merge-with + acc (fingerprint-table! table))]
-                   (if (and (continue? new-acc) (not (sync-util/abandon-sync? new-acc)))
+                 (let [ret (fingerprint-table! table)
+                       new-acc (let [ret (if (:throwable ret)
+                                           (-> ret
+                                               (dissoc :throwable)
+                                               (update :failed-fingerprints inc))
+                                           ret)]
+                                 (merge-with + acc ret))]
+                   (if (and (continue? ret) (not (sync-util/abandon-sync? ret)))
                      new-acc
                      (reduced new-acc))))
                (empty-stats-map 0)

--- a/test/metabase/sync/analyze/fingerprint_test.clj
+++ b/test/metabase/sync/analyze/fingerprint_test.clj
@@ -347,7 +347,7 @@
                                                              (swap! tables-fingerprinted inc)
                                                              (throw (CannotAcquireResourceException. "Unable to acquire resource")))
                       sync.fingerprint/*refingerprint?* true]
-          (is (=? (assoc (sync.fingerprint/empty-stats-map 0)
-                         :throwable #(instance? CannotAcquireResourceException %))
-                  (sync.fingerprint/fingerprint-fields-for-db! (mt/db) (constantly nil))))
+          (is (= (assoc (sync.fingerprint/empty-stats-map 0)
+                        :failed-fingerprints 1)
+                 (sync.fingerprint/fingerprint-fields-for-db! (mt/db) (constantly nil))))
           (is (= 1 @tables-fingerprinted)))))))

